Initial Conditions may depend on UserObjects

permcody opened the pull request:
Funny, we had the UserObject interface coupled into ICs before, but that doesn't mean that they'd execute before the initial condition. This branch adds the ability to execute UOs _before_ the initial conditions are projected so we can run more complicated algorithms before the ICs are set up. A small test is included but this PR is the precursor to a new system (#8810) that we'll use extensively in the phase-field module.

permcody committed:
Allow UserObjects to execute before ICs

refs #8810

permcody committed:
Adding test of depend IC coupling

refs #8810

aeslaughter commented on framework/include/ics/InitialCondition.h:
```diff
@@ -95,6 +97,13 @@ class InitialCondition : public MooseObject,
   virtual const std::set<std::string> & getSuppliedItems() override;
 
+  template <typename T>
+  const T & getUserObject(const std::string & name);
```
override keyword?

aeslaughter commented on framework/include/ics/InitialCondition.h:
```diff
@@ -95,6 +97,13 @@ class InitialCondition : public MooseObject,
+  template <typename T>
+  const T & getUserObject(const std::string & name);
+  template <typename T>
+  const T & getUserObjectByName(const UserObjectName & name);
```
override?

aeslaughter commented on framework/include/ics/InitialCondition.h:
```diff
@@ -95,6 +97,13 @@ class InitialCondition : public MooseObject,
+  template <typename T>
+  const T & getUserObjectByName(const UserObjectName & name);
+
+  const UserObject & getUserObjectBase(const std::string & name);
```
override?

aeslaughter reviewed (dismissed):
Fine with me.

aeslaughter commented:
To answer my own question, override doesn't work for template functions, because they can't be virtual.

moosebuild commented:
Job `App tests:linux-gnu` on dfa3b77 : invalidated by @permcody

BISON PR Merged

permcody commented:
@YaqiWang - This PR may be of interest to you. It adds native capability to MOOSE to allow IC -> UO dependencies. Unfortunately, it totally breaks Rattlesnake right now due to your custom SPH Executioner. Hopefully we can work that out.

friedmud commented:
This looks good permcody !
On Tue, Apr 18, 2017 at 8:17 AM Cody Permann <notifications@github.com>
wrote:

> @YaqiWang <https://github.com/YaqiWang> - This PR may be of interest to
> you. It adds native capability to MOOSE to allow IC -> UO dependencies.
> Unfortunately, it totally breaks Rattlesnake right now due to your custom
> SPH Executioner. Hopefully we can work that out.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8945#issuecomment-294622917>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMXsyFuKVI_YC0ACBfzJfctq6gjTKks5rw_MkgaJpZM4M_k61>
> .
>

permcody committed:
Hacking up AuxGroupExecutMooseObjectWarehouse

This PR only puts objects in the PRC_IC group that execute on INITIAL.
If an object has more than one execute_on flag. It will be placed in the
PRE/POST aux groups for the remaining flags.

YaqiWang commented:
Yes just looking at the title, I was wondering if I can remove the hack in our SPH executioner. It could be more complicated, I will talk with you when I am back.

permcody commented:
@YaqiWang or @snschune, How do you want me to handle these changes? This PR properly runs UOs with `execute_on = initial` before any IC that fetches that UO. I looked briefly at the failed test and it looks like the hacked up Executioner is basically doing this manually. I may need some help untangling it. Is it possible to dump that Executioner, or does it do more than just run UOs before ICs? (I haven't really looked at it in depth)

YaqiWang commented:
Sorry, still trying to pick up things back from the conference. I remember I had to do the hack because of potential cyclic dependency of IC and UO, i.e. the UO depends on a variable, which could be set by a IC and later another IC depends on the UO. I will grab this branch and test it out today.

permcody commented:
Oh, I may not be able to handle the cyclic dependency if your UO depends on a variable but is also depended upon by another IC. Yuck

YaqiWang commented:
@permcody the simplest solution is to bypass the dependency setup in the SPHIC by calling `UserObjectInterface::getUserObject` instead of `getUserObject`, which is changed in this PR. Because this is your PR, would you mind to send in a small patch in Rattlesnake?

But your PR has a problem: sometimes ICs requires a user object, whose initial setup is sufficient for the ICs. But you assume here a depending UO must finish its execution for the IC. This cause the failure of some of Rattlesnake SPH tests. IC can call `UserObjectInterface::getUserObject` of cause to bypass this but I think you may want to re-think the default behavior otherwise users could be surprised. Some other SPH test failures are caused by the IC-to-UO dependency (the SPH UO depends on IC but used by setting IC).

permcody committed:
Allow users to override automatic UO dependency resolution

permcody commented:
@YaqiWang - I added one more commit that adds a flag to the IC base class to ignore the UO dependency resolution. I think that this should be a temporary solution. The good news it only requires a single change to Rattlesnake. I have a patch ready.

moosebuild commented:
Job `App tests:linux-gnu` on 337ab39 : invalidated by @permcody

Rattlesnake updated
